core/src/main/java/ch/cyberduck/core/Resolver.java

```diff
@@ -67,20 +68,25 @@ public Resolver(final boolean preferIPv6) {
      */
     public InetAddress[] resolve(final String hostname, final CancelCallback callback) throws ResolveFailedException, ResolveCanceledException {
         final CountDownLatch signal = new CountDownLatch(1);
-        final AtomicReference<Set<InetAddress>> resolved = new AtomicReference<>();
+        final AtomicReference<List<InetAddress>> resolved = new AtomicReference<>();
         final AtomicReference<UnknownHostException> failure = new AtomicReference<>();
         final Thread resolver = threadFactory.newThread(new Runnable() {
             @Override
             public void run() {
                 try {
                     final InetAddress[] allByName = InetAddress.getAllByName(hostname);
-                    resolved.set(Arrays.stream(allByName).collect(Collectors.toSet()));
-                    if(preferIPv6) {
-                        final Set<InetAddress> filtered = Arrays.stream(allByName).filter(inetAddress -> inetAddress instanceof Inet6Address).collect(Collectors.toSet());
-                        if(!filtered.isEmpty()) {
-                            resolved.set(filtered);
+                    resolved.set(Arrays.stream(allByName).sorted(new Comparator<InetAddress>() {
+                        @Override
+                        public int compare(final InetAddress o1, final InetAddress o2) {
+                            if(o1 instanceof Inet6Address) {
```

Comment:
Order of IPv6 addresses as in `allByName` is reversed when `preferIPv6` is `true`. Should be `o1 instanceof Inet6Address && o2 instanceof Inet4Address`. Same for the following test.